spec/reek/configuration/app_configuration_spec.rb

```diff
@@ -13,72 +14,61 @@
-      described_class.configure_smell_repository smell_repository
+    it 'properly loads configuration and processes it' do
+      finder = Reek::Configuration::ConfigurationFileFinder
+      allow(finder).to receive(:find_by_cli).and_return full_configuration_path
 
-      expect(smell_repository.detectors[Reek::Smells::DataClump]).to be_enabled
-      expect(smell_repository.detectors[Reek::Smells::UncommunicativeVariableName]).
-        not_to be_enabled
-      expect(smell_repository.detectors[Reek::Smells::UncommunicativeMethodName]).not_to be_enabled
+      expect(subject.exclude_paths).to eq(expected_exclude_paths)
+      expect(subject.default_directive).to eq(expected_default_directive)
+      expect(subject.directory_directives).to eq(expected_directory_directives)
     end
   end
 
-  describe '.exclude_paths' do
-    let(:config_path) do
-      SAMPLES_PATH.join('configuration/with_excluded_paths.reek')
-    end
+  describe '#directory_directive_for_source' do
+    context 'returns something falsey' do
+      it 'for sources from IO' do
+        expect(subject.directory_directive_for_source(Reek::Source::SourceCode::IO_IDENTIFIER)).
+          to be_falsey
+      end
 
-    it 'should return all paths to exclude' do
-      with_test_config(config_path) do
-        expect(described_class.exclude_paths).to eq [
-          SAMPLES_PATH.join('source_with_exclude_paths/ignore_me'),
-          SAMPLES_PATH.join('source_with_exclude_paths/nested/ignore_me_as_well')
-        ]
+      it 'for sources from strings' do
+        expect(subject.directory_directive_for_source(Reek::Source::SourceCode::STRING_IDENTIFIER)).
+          to be_falsey
       end
     end
-  end
 
-  describe '.load_from_file' do
-    it 'loads the configuration from given file' do
-      described_class.load_from_file(sample_configuration_path)
-      expect(described_class.configuration).to eq(sample_configuration_loaded)
-    end
-  end
+    context 'our source is in a directory for which we have a directive' do
+      let(:baz_config)  { { Reek::Smells::IrresponsibleModule => { enabled: false } } }
+      let(:bang_config) { { Reek::Smells::Attribute => { enabled: true } } }
 
-  describe '.reset' do
-    it 'resets the configuration' do
-      described_class.load_from_file(sample_configuration_path)
+      let(:directory_directives) do
+        {
+          Pathname.new('foo/bar/baz')  => baz_config,
+          Pathname.new('foo/bar/bang') => bang_config
+        }
+      end
+      let(:source_via) { 'foo/bar/bang/dummy.rb' }
 
-      expect(described_class.configuration).to eq(sample_configuration_loaded)
-      described_class.reset
-      expect(described_class.configuration).to eq(default_configuration)
-      expect(described_class.exclude_paths).to be_empty
+      it 'returns the corresponding directive' do
+        allow(subject).to receive(:directory_directives).and_return directory_directives
```

Comment:
What do you mean? `directory_directives=` is private and the initializer for AppConfiguration requires the unholy option struct with a pathname. I'm not against changing this code to be more flexible, I'd love it if we 1.) refactored the options hell and 2.) made app_configuration so flexible that we can pass in different things and it still works out.
However, I'd like to do that in a follow-up pull request, since this is pretty big. For [1] there is already an issue. I'll open up another for [2].